Add wxImage ChangeSaturation(), ChangeBrightness(), ChangeHSV() and C…

tomay3000 opened the pull request:
…hangeLightness() functions

tomay3000 committed:
Add wxImage ChangeSaturation(), ChangeBrightness(), ChangeHSV() and ChangeLightness() functions

tomay3000 commented:
Reserved

vadz reviewed (commented):
Thanks, this definitely looks useful!

However I'm indeed not totally happy about the code duplication between all these different functions. I wonder if we could provide a generic `ApplyToAllPixels(RGBValue (*func)(RGBValue rgb))` and implement them all (including the existing `ConvertToDisabled()` as simple wrappers around it? This would save quite some code, provide more flexibility and would IMO make things more clear. Do you think you could redo this PR like this?

Also, it would be nice to include your patch 2 into the image sample, as a new frame there. Could you please add it to the PR?

And, last, but not least, could you please at least some very simple tests to `tests/image/image.cpp` to check that these functions behave as expected?

TIA!

tomay3000 commented:
> However I'm indeed not totally happy about the code duplication between all these different functions. I wonder if we could provide a generic `ApplyToAllPixels(RGBValue (*func)(RGBValue rgb))` and implement them all (including the existing `ConvertToDisabled()` as simple wrappers around it? This would save quite some code, provide more flexibility and would IMO make things more clear.

You are absolutely right, I have seen that in the first place and I was wondering how to optimize the code to be using one single function with an arbitrary function as parameter to do the `loop` RGBValue modifications. and I guess, yes, it should be done this way.

> Do you think you could redo this PR like this?

Sure, I am gonna do it.

> Also, it would be nice to include your patch 2 into the image sample, as a new frame there. Could you please add it to the PR?

Sure.
 
> And, last, but not least, could you please at least some very simple tests to `tests/image/image.cpp` to check that these functions behave as expected?

Yes.

tomay3000 commented on interface/wx/image.h:
```diff
@@ -1017,6 +1046,16 @@ class wxImage : public wxObject
+        This utility function simply darkens or lightens a color, based on the
+        specified percentage @a ialpha. @a ialpha of 0 would make the color
+        completely black, 200 completely white and 100 would not change the color.
+
```
Ok.

tomay3000 commented on src/common/image.cpp:
```diff
@@ -1852,6 +1852,52 @@ wxImage wxImage::ConvertToDisabled(unsigned char brightness) const
+    const int height = M_IMGDATA->m_height;
+    const int nSize = width * height;
+    image.Create(width, height, false);
+    const unsigned char *pByAlpha = M_IMGDATA->m_alpha;
```
> we generally never use Hungarian notation.

Good to know, I will keep this in mind in future pull-requests.

tomay3000 commented on src/common/image.cpp:
```diff
@@ -3273,6 +3319,145 @@ void wxImage::RotateHue(double angle)
+void wxImage::ChangeSaturation(double factor)
+{
+    wxASSERT (factor >= -1.0 && factor <= 1.0);
+    AllocExclusive();
```
I guess I think so.

tomay3000 committed:
Refactor the code

 is adopted to do the loop modifications instead of repeating it in each function.

tomay3000 commented:
`template <typename T> void ApplyToAllPixels(void (wxImage::*func)(unsigned char *, T), T value);`
 is adopted to do the loop modifications instead of repeating it in each function.

tomay3000 committed:
Update the documentation.
Mention that each of `ConvertToGreyscale()`, `ConvertToMono()`, `ConvertToDisabled()` and `ChangeLightness()` functions
are respectively calling `wxColour::MakeGrey()`, `wxColour::MakeMono()`, wxColour::MakeDisabled()` and `wxColour::ChangeLightness()` for each pixel in the image.

tomay3000 committed:
Improve the code
Just reference-count the image instead of making a copy of it by calling wxImage::Copy(), because it is gonna be cloned anyway later in the call of wxImage::ApplyToAllPixels() by calling wxImage::AllocExclusive().

tomay3000 committed:
Update the documentation.
Upgrade the version of the added functions to 3.1.6 after 3.1.5 release.

vadz commented on src/common/image.cpp:
```diff
@@ -3232,45 +3187,126 @@ wxImage::RGBValue wxImage::HSVtoRGB(const HSVValue& hsv)
+    RGBValue rgbValue(rgb[0], rgb[1], rgb[2]);
+    HSVValue hsvValue = RGBtoHSV(rgbValue);
+
+        hsvValue.value += hsvValue.value * factor;
+
+        if (hsvValue.value > 1.0)
+            hsvValue.value = 1.0;
+        else if (hsvValue.value < 0.0)
+            hsvValue.value = 0.0;
```
Please check/fix the indentation here.

tomay3000 replied:
Oops.

vadz commented on include/wx/image.h:
```diff
@@ -628,6 +651,46 @@ class WXDLLIMPEXP_CORE wxImage: public wxObject
+    // Makes an RGB value disabled.
+    void DoMakeDisabled(unsigned char *rgb, unsigned char brightness);
+
+    // red, green and blue are weight doubles, they are used internally by
```
Does "weight" mean "in [0..1] range"? If so, it'd be useful to mention this in the comment explicitly.

tomay3000 replied:
Exactly.

vadz commented on include/wx/image.h:
```diff
@@ -628,6 +651,46 @@ class WXDLLIMPEXP_CORE wxImage: public wxObject
 
+    // Helper function used internally by wxImage class only.
+    template <typename T>
+    void ApplyToAllPixels(void (wxImage::*func)(unsigned char *, T), T value);
```
It seems to me that none of the functions used with this one actually needs to be a member function, so I'd use
```suggestion
    void ApplyToAllPixels(void (*filter)(const RGBValue* mask, unsigned char *, T), T value);
```
here and just pass it the mask RGB (the pointer would be null if there is no mask).

I suspect that it would be beneficial to have 2 different execution paths/loops for the case with/without mask as the latter should be quite a bit faster as we don't need any tests at all inside the loop, and so it should be auto-vectorized, but this would require more changes.

tomay3000 replied:
I'll move them to the source file, and maybe make `ApplyToAllPixels` `protected` to allow sub-classing customization.
the mask condition check `if ( !HasMask() || rgb[0] != GetMaskRed() || rgb[1] != GetMaskGreen() || rgb[2] != GetMaskBlue() )` is called for each pixel and is composed from **OR**s.

vadz commented on include/wx/image.h:
```diff
@@ -628,6 +651,46 @@ class WXDLLIMPEXP_CORE wxImage: public wxObject
+
+    // red, green and blue are weight doubles, they are used internally by
+    // DoMakeGrey() only.
+    class WeightValue
```
If we don't need `DoXXX()` to be member functions, they, and this class, should be moved to `image.cpp`.

tomay3000 replied:
Absolutely, but in an unnamed namespace or each one just before its caller function?

vadz reviewed (commented):
Thanks a lot! The changes look good, but might be improved a bit more and, especially, I'd still very much like to have at least some trivial tests for this.

Thanks again!

tomay3000 commented:
> Thanks a lot! The changes look good, but might be improved a bit more and, especially, I'd still very much like to have at least some trivial tests for this.
> 
> Thanks again!

Thank you,
My daily work kept me from finishing the sample frame, I left it at 90%, I will finish it tonight.

tomay3000 committed:
Fix a wrong indentation.

tomay3000 committed:
Mention the range of the private class WeightValue member variables.

tomay3000 committed:
Move DoXXX() functions and WeightValue class to the source file image.cpp

tomay3000 committed:
Make ApplyToAllPixels protected to allow more subclassing filters.

tomay3000 committed:
Add a new filters test frame to samples/image/image.cpp

tomay3000 commented:
I think that's it, except for the tests which I don't know what convention is adopted for them especially for this kind of objects!

vadz commented:
Thanks, I'll look at this a.s.a.p. but the tests would be just some very simple functions in `tests/image/image.cpp`, e.g.
```cpp
TEST_CASE("wxImage::ChangeColours", "[image]")
{
   ... define the XPMs with the correct data ... 
   wxImage image(normal_xpm); // Or create it in some other way, e.g. make a 2*2 image by directly initializing its pixels.
   wxImage imageSaturated(saturated_xpm);
   CHECK_THAT( image.ChangeSaturation(), RGBSameAs(imageSaturated) ); // Or test individual pixels.
   ... same for the other functions ...
}
```
This could be as simple as you can make it, but would at least provide a modicum of assurance that this works as expected. Do you think you could please add such a test? TIA!

Note to self: clang CI failure is bogus (but concerning...).

tomay3000 committed:
Added some tests.
New Catch tests have been added for each of RotateHue(), ChangeSaturation(), ChangeBrightness(), ChangeHSV(), ChangeLightness(), ConvertToDisabled(), ConvertToGreyscale() and ConvertToMono() functions.

tomay3000 commented:
I didn't notice the test transition to `Catch` also the dedicated function: `RGBSameAs()`. I have to catch-up to the new code. I also found a lot of conflicts about the last commits which became messy, because my repo is only `3.1.3`. I have to update my repo to use the latest changes available without losing these commit changes.

tomay3000 commented:
I don't know why the tests are failing.
Maybe because the png images have not been uploaded!

vadz commented:
I think it's just because the tests run from the `tests` directory and your files are under `tests/image`, isn't it?

But then I don't really understand how could they pass even locally -- do they actually pass for you?

In any case, it would be better to use `REQUIRE()` around `LoadFile()` calls as if it fails, it's useless to continue with the test case.

tomay3000 commented:
Indeed, the reason that it passed for me locally before is because I run it from `tests/image` directory.
I have preceded the file names with `"image/"` and the test passes now.

vadz commented:
Not sure if you've noticed it, but there still seems to be an [error in CMake build](https://ci.appveyor.com/project/wxWidgets/wxwidgets/builds/38886634/job/vug9hvh8ru02x7ip#L3363).

tomay3000 commented:
> Not sure if you've noticed it, but there still seems to be an [error in CMake build](https://ci.appveyor.com/project/wxWidgets/wxwidgets/builds/38886634/job/vug9hvh8ru02x7ip#L3363).

Well, that's really annoying.
I think the problem comes from `appveyor-test.bat` in `cmake` label.

vadz commented:
Sorry, I really don't have time to look at this yet, but I just wonder if you could do the same thing we already do in the other tests using these files -- they do pass in all CI builds, so what's the difference with the new tests?

tomay3000 committed:
Added missing image file names for cmake GUI test.

MaartenBent commented:
You need to add the images to `build/cmake/tests/gui/CMakeLists.txt`, same as you did for `test.bkl`.

tomay3000 commented:
> You need to add the images to `build/cmake/tests/gui/CMakeLists.txt`, same as you did for `test.bkl`.

Exactly, I just found it too while I was digging a bit more.
The reason I didn't see it the first time is because `cmake` test files are stored in the `build` directory.

Thank you.

vadz commented:
Great, thanks a lot! I'm (squash) merging this now and will push soon.
